multiqc/utils/config.py

```diff
@@ -62,9 +59,32 @@
 # Modules must be listed in setup.py under entry_points['multiqc.modules.v1']
 # Get all modules, including those from other extension packages
 avail_modules = dict()
+module_roots = set()
 for entry_point in pkg_resources.iter_entry_points('multiqc.modules.v1'):
-    nicename = str(entry_point).split('=')[0].strip()
-    avail_modules[nicename] = entry_point
+    avail_modules[entry_point.name] = entry_point
+
+    # Infer where the module is actually going to be loaded from.
+    module_roots.add(os.path.join(entry_point.dist.location, entry_point.module_name.split(".")[0]))
+
+# Module filename search patterns
+# Look for search_patterns.yaml in the utils directory where the module is located,
+# so that plugin modules can add their own patterns.
+sp = dict()
+for searchp_fn in [ os.path.join( d, 'utils', 'search_patterns.yaml') for d in module_roots ]:
+    try:
+        with open(searchp_fn) as f:
+            sp.update(yaml.load(f))
+    except OSError:
+        pass
+
+# We can't log yet, but this is a serious error.
+if not sp:
+    print("No default search patterns were loaded. Looked in locations:\n{}".format(pformat(module_roots)),
```

Comment:
Would this ever happen? Surely the base config is always going to load with some search patterns?

Reply:
I guess this is more of an assertion failure than a runtime error since every module should come with a matching set of file patterns to scan.
If the module is defined as part of a plugin (which is what I'm working on) then the main base config has no idea about the module and what files to feed it. To me, having all the plugins listed out in a monolithic base config file goes against the modular ethos anyway, so this is my attempt to fix it both for plugins and in general by allowing config fragments to be discovered and merged.

Reply:
Ah I see, sorry I misunderstood this change and assumed that this was an extra option to allow a `search_patterns.yaml` file for external modules.

Yes, I don't have a problem with breaking up the central `search_patterns.yaml` file into smaller files that are in each module directory. They used to be kept in the main `config_defaults.yaml` file, which in turn used to all be hardcoded into `config.py` (which in turn used to be part of `multiqc.py` a long, long time ago)...